internal/envelope/conn/envelope_conn.go

```diff
@@ -131,20 +140,41 @@ func (e *EnvelopeConn) Serve() error {
 	// goroutine for every request because we must guarantee that requests are
 	// processed in order. Logs, for example, need to be received and processed
 	// in order.
-	group.Go(func() error {
+	e.running.Go(func() error {
 		for {
-			msg, ok := <-msgs
-			if !ok {
-				return nil
-			}
-			if err := e.handleMessage(msg); err != nil {
-				e.conn.cleanup(err)
-				return err
+			select {
+			case msg := <-msgs:
+				if err := e.handleMessage(msg, h); err != nil {
+					e.stop(err)
+					return err
+				}
+			case <-e.ctx.Done():
+				e.stop(e.ctx.Err())
+				return e.ctx.Err()
 			}
 		}
 	})
+}
+
+// Wait waits for the connection to terminate. It returns an error that
```

Comment:
We chatted offline and might merge `Wait` into `Serve`. I think that will make `EnvelopeConn` consistent with `WeaveletConn` as well.